packages/typescript/lib/quickstart/createLanguageServicePlugin.ts

```diff
@@ -1,114 +1,42 @@
-import { FileMap, Language, LanguagePlugin, createLanguage } from '@volar/language-core';
 import type * as ts from 'typescript';
-import { resolveFileLanguageId } from '../common';
 import { createProxyLanguageService } from '../node/proxyLanguageService';
-import { decorateLanguageServiceHost, searchExternalFiles } from '../node/decorateLanguageServiceHost';
-
-export const externalFiles = new WeakMap<ts.server.Project, string[]>();
-export const projectExternalFileExtensions = new WeakMap<ts.server.Project, string[]>();
-export const decoratedLanguageServices = new WeakSet<ts.LanguageService>();
-export const decoratedLanguageServiceHosts = new WeakSet<ts.LanguageServiceHost>();
+import { createLanguageCommon, isHasAlreadyDecoratedLanguageService, makeGetExternalFiles, projectExternalFileExtensions } from './languageServicePluginCommon';
+import type { createPluginCallbackSync } from './languageServicePluginCommon';
 
+/**
+ * Creates and returns a TS Service Plugin using Volar primitives.
+ *
+ * See https://github.com/microsoft/TypeScript/wiki/Writing-a-Language-Service-Plugin for
+ * more information.
+ */
 export function createLanguageServicePlugin(
-	create: (
-		ts: typeof import('typescript'),
-		info: ts.server.PluginCreateInfo
-	) => {
-		languagePlugins: LanguagePlugin<string>[],
-		setup?: (language: Language<string>) => void;
-	}
+	createPluginCallback: createPluginCallbackSync
 ): ts.server.PluginModuleFactory {
 	return modules => {
 		const { typescript: ts } = modules;
+
 		const pluginModule: ts.server.PluginModule = {
 			create(info) {
-				if (
-					!decoratedLanguageServices.has(info.languageService)
-					&& !decoratedLanguageServiceHosts.has(info.languageServiceHost)
-				) {
-					decoratedLanguageServices.add(info.languageService);
-					decoratedLanguageServiceHosts.add(info.languageServiceHost);
-
-					const { languagePlugins, setup } = create(ts, info);
-					const extensions = languagePlugins
+				if (!isHasAlreadyDecoratedLanguageService(info)) {
+					const createPluginResult = createPluginCallback(ts, info);
+					const extensions = createPluginResult.languagePlugins
 						.map(plugin => plugin.typescript?.extraFileExtensions.map(ext => '.' + ext.extension) ?? [])
 						.flat();
+
+					// TODO: this logic does not seem to appear in the async variant
+					// (createAsyncLanguageServicePlugin)... bug?
```

Comment:
OPEN QUESTION: is it a bug that the async variant (`createAsyncLanguageServicePlugin`) does not add the extensions passed in to `projectExternalFileExtensions`? (This is also how it was prior to this PR).

Reply:
This is intentional, in the case of async this will cause some internal error in the tsserver, which is why we need to explicitly pass the required extensions in `createAsyncLanguageServicePlugin` parameters.